Fix flicker when selecting tiles with custom dimensions

nfmohit opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #7742 

## Relevant technical choices

This PR aims to resolve the reported flicker in #7742 when a tile is selected that depends on unavailable custom dimensions, by not closing the panel and showing a loader in CTA when the user is being redirected to OAuth.

### Deviation from IB

The proposed IB works, however, since the open state of the metrics selection panel is tracked using a key in the `core/ui` store and the `setPermissionScopeError` action/`AuthenticatedPermissionsModal` component unconditionally snapshots the stores, when the user returns from the OAuth flow, the metrics selection panel is reopened.

To address this, this PR removes the usage of the `setPermissionScopeError` action in this regard, and introduces a baked-in approach to snapshot only the `core/forms` store (which is required) and redirect to the OAuth URL.

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

zutigrm commented on assets/js/components/KeyMetrics/MetricsSelectionPanel/Footer.js:
```diff
@@ -144,47 +162,42 @@ export default function Footer( { savedMetrics } ) {
-						} );
+					if ( ! hasAnalytics4EditScope && OAuthURL ) {
+						snapshotStore( CORE_FORMS, registry );
+						navigateTo( OAuthURL );
```
Maybe we can introduce new property in the `setPermissionScopeError` object, like `snapshotStore` that would accept store name, in this case `CORE_FORMS`, and then let `AuthenticatedPermissionsModal` handle this in centralised manner, by [checking](https://github.com/google/site-kit-wp/blob/ddbc12a5f33a76d6feaae8451d72a91d958e89ac/assets/js/components/PermissionsModal/AuthenticatedPermissionsModal.js#L62)  if `permissionsError?.snapshotStore`, to snapshot only passed custom store name, otherwise fallback to default snapshot of all stores that implement that method, like it is doing now.

This way it would be less "hacky" and no need to include all these dependencies like `registry`, and `snapshotStore` just for this, and it will be easier to use it in the future by sticking to the common pattern. It also requires less code.

nfmohit replied:
Updated, thank you!

zutigrm reviewed (commented):
@nfmohit Thanks, looks good, snapshotting only single (required) store per need might be useful in the future if similar behaviour of skipping the snapshots of other stores is needed.

I just have 1 comment/suggestion.

Let me know what you think

nfmohit commented:
> @nfmohit Thanks, looks good, snapshotting only single (required) store per need might be useful in the future if similar behaviour of skipping the snapshots of other stores is needed.
> 
> I just have 1 comment/suggestion.
> 
> Let me know what you think

Thank you for the kind review, @zutigrm. I've addressed your feedback. Thanks!

zutigrm reviewed (approved):
Thank you @nfmohit looks good. 

I just quickly removed optional chaining from snapshot store name within the condition, since it is already verified that value is not `null`/`undefined` in the condition above.

eugene-manuilov commented:
Thanks, @nfmohit. Overall it looks solid but I am a bit hesitant of introducing changes for snapshots in this PR. Maybe instead of changing how snapshots work, we should consider the `SideSheet` element in the `MetricsSelectionPanel` component when we have that key in the UI datastore **or we are navigating to an OAuth page**? What do you think? This seems like a more simple change than snapshots modifications.

nfmohit commented:
> Thanks, @nfmohit. Overall it looks solid but I am a bit hesitant of introducing changes for snapshots in this PR. Maybe instead of changing how snapshots work, we should consider the `SideSheet` element in the `MetricsSelectionPanel` component is still open when we have that key in the UI datastore **or we are navigating to an OAuth page**? What do you think? This seems like a more simple change than snapshots modifications.

Thank you @eugene-manuilov. I had actually explored this approach during execution, but there is a concern about the `SideSheet` opening when any other similar redirections to OAuth happen. This is less of a concern now I think since [#7758](https://github.com/google/site-kit-wp/issues/7758) changed the redirect URL for the OAuth URL, however, the same redirect URL is also used within `withCustomDimensions`. Hence, I don't think it is a good idea to use the built-in `isNavigating` selector to achieve this.

However, alternatively, we can use a local state within the component to track if we're redirecting the user, and use that to keep the `SideSheet` open. Do you suggest we go that route?

nfmohit commented:
> Ok, let's use the local state then. Or if we can let's lift it up to the parent component.

Updated accordingly, let me know if it looks good, thank you @eugene-manuilov !

eugene-manuilov reviewed (approved):
Thanks, @nfmohit. Looks good to me now.
